src/components/SuggestionBox.tsx

```diff
@@ -7,10 +8,21 @@ import { QueryInput } from './SuggestionBox/QueryInput';
 export type Item<T> = { element: ComponentChild; searchableText: string; value: T };
 
 export function matchItems<T>(query: string, items: Item<T>[]): Item<T>[] {
-  return items.filter((item) => {
-    const target = item.searchableText.toLowerCase();
-    return target.includes(query.toLowerCase());
-  });
+  const maxAambig = Math.min(Math.floor(query.length / 4) + 1, 3);
+  const match = Asearch(` ${query} `); // 部分一致できるように、両端をスペースで囲む
+  // あいまい度の少ない項目から順に並べる
+  // 重複は除く
+  const indice: Set<number> = new Set();
+  const result: Item<T>[] = [];
+  for (let ambig = 0; ambig <= maxAambig; ambig++) {
+    items.forEach((item, index) => {
+      if (!match(item.searchableText, ambig)) return;
+      if (indice.has(index)) return;
+      indice.add(index);
+      result.push(item);
+    });
+  }
```

Comment:
MAY: index わざわざ使わずとも、以下のように書けば分かりやすいかなと思っているのですがいかがでしょうか。

```suggestion
  const pushedItems: Set<Item> = new Set();
  const result: Item<T>[] = [];
  for (let ambig = 0; ambig <= maxAambig; ambig++) {
    items.forEach((item) => {
      if (!match(item.searchableText, ambig)) return;
      if (pushedItems.has(item)) return;
      pushedItems.add(item);
      result.push(item);
    });
  }
```

Reply:
あれ？`Set`による重複除去って、任意のobjectで有効でしたっけ？

Reply:
試してみました。やはり重複除去はされなさそうです。
https://runkit.com/takker99/test-js-set
[![Image from Gyazo](https://i.gyazo.com/a0130af2d8ec3040fd50e2b0a792e01f.png)](https://gyazo.com/a0130af2d8ec3040fd50e2b0a792e01f)

Reply:
それは `{foo:45,bar:77}` とオブジェクトリテラルを使って新しいオブジェクトを作り直しているからですね。それぞれ別々のオブジェクトなので別々のものとして扱われます。

こんな感じで、オブジェクトリテラルで毎回新しいオブジェクトを生成せずに、一度生成したものを使い回せばちゃんと重複削除できます。

```js
const obj = { foo: 45, bar: 77 };
const set = new Set([obj, obj]);
console.log(Array.from(set)) // [{ foo: 45, bar: 77 }]
```

Reply:
例:
```js
console.log({} === {}); // false

const obj = {};
console.log(obj === obj); // true
```

Reply:
本当だ！教えていただきありがとうございます！

C++でいう変数のアドレス比較みたいな感じですね。なるほど。